test/framework/finalizers_helpers.go

```diff
@@ -0,0 +1,168 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterctlcluster "sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// finalizerAssertion contains a list of expected Finalizers corresponding to a resource Kind.
+var finalizerAssertion = map[string][]string{
+	"Cluster":             {clusterv1.ClusterFinalizer},
+	"Machine":             {clusterv1.MachineFinalizer},
+	"MachineSet":          {clusterv1.MachineSetTopologyFinalizer},
+	"MachineDeployment":   {clusterv1.MachineDeploymentTopologyFinalizer},
+	"ClusterResourceSet":  {addonsv1.ClusterResourceSetFinalizer},
+	"DockerMachine":       {infrav1.MachineFinalizer},
+	"DockerCluster":       {infrav1.ClusterFinalizer},
+	"KubeadmControlPlane": {controlplanev1.KubeadmControlPlaneFinalizer},
+}
+
+// ValidateFinalizersResilience checks that expected finalizers are in place, deletes them, and verifies that expected finalizers are properly added again.
+func ValidateFinalizersResilience(ctx context.Context, proxy ClusterProxy, namespace, clusterName string) {
+	clusterKey := client.ObjectKey{Namespace: namespace, Name: clusterName}
+
+	// Collect all objects where finalizers were initially set
+	objectsWithFinalizers := getObjectsWithFinalizers(ctx, proxy, namespace)
+
+	// Removes all the Finalizers.
+	// We are testing the worst-case scenario, i.e. all finalizers are deleted.
+	// The reconciliation loop should be able to recover from this, by adding the required Finalizers back.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, true)
+
+	// Once all Clusters are paused remove the Finalizers from all objects in the graph.
+	removeFinalizers(ctx, proxy, namespace)
+
+	// Unpause the cluster.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, false)
+
+	// Annotate the ClusterClass, if one is in use, to speed up reconciliation. This ensures ClusterClass Finalizers
+	// are re-reconciled before asserting the owner reference graph.
+	forceClusterClassReconcile(ctx, proxy.GetClient(), clusterKey)
```

Comment:
Do we have finalizers related to ClusterClass?

Reply:
Yup, but the corresponding controllers do not seem to watch on ClusterClass. So I suspect forceClusterClass reconcile doesn't do anything.

Ah maybe the goal is that the ClusterClass reconciler itself is run and we want to make sure that this doesn't somehow break the finalizers. Although this reconciler doesn't have anything to do with the MD & MS objects (?)

Reply:
Your assumption is right if I understood it correctly.

Reply:
Then I would drop it if we don't depend on it here

Reply:
It does speed-up reconciliations for topology Clusters https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/topology/cluster/cluster_controller.go#L101

Reply:
Yup but this reconciler is not setting any of our finalizers. To be honest. I'm fine either way, it's just a detail

Reply:
I will remove it  - tested it locally, do not a significant difference in performance.

Reply:
Thx!